virtual/registry/kv_registry.go

```diff
@@ -221,149 +221,100 @@ func (k *kvRegistry) EnsureActivation(
 	req EnsureActivationRequest,
 ) (EnsureActivationResult, error) {
 	actorKey := getActorKey(req.Namespace, req.ActorID, req.ModuleID)
-	references, err := k.kv.Transact(func(tr kv.Transaction) (any, error) {
-		ra, ok, err := k.getActor(ctx, tr, actorKey)
-		if err == nil && !ok {
-			_, err := k.createActor(
-				ctx, tr, req.Namespace, req.ActorID, req.ModuleID, types.ActorOptions{})
-			if err != nil {
-				return nil, fmt.Errorf("EnsureActivation: error creating actor: %w", err)
-			}
-			ra, ok, err = k.getActor(ctx, tr, actorKey)
-			if err != nil {
-				return nil, fmt.Errorf("EnsureActivation: error getting actor: %w", err)
-			}
-			if !ok {
-				return nil, fmt.Errorf(
-					"[invariant violated] error ensuring activation of actor with ID: %s, does not exist in namespace: %s, err: %w",
-					req.ActorID, req.Namespace, errActorDoesNotExist)
-			}
-		}
-		if err != nil {
-			return nil, fmt.Errorf("EnsureActivation: error getting actor: %w", err)
-		}
-		if !ok {
-			// Make sure we use %w to wrap the errActorDoesNotExist so the caller can use
-			// errors.Is() on it.
-			return nil, fmt.Errorf(
-				"[invariant violated] error ensuring activation of actor with ID: %s, does not exist in namespace: %s, err: %w",
-				req.ActorID, req.Namespace, errActorDoesNotExist)
-		}
 
-		serverKey := getServerKey(ra.Activation.ServerID)
-		v, ok, err := tr.Get(ctx, serverKey)
+	// Perform a transaction to ensure atomicity.
```

Comment:
Very random but another issue we should probably file in the repo is that the registry does not "remember" blacklisted servers. It only knows the single blacklisted server id the caller tells it. We should probably make the registry accumulate a set of known blacklisted servers for each actor (with a blacklist end time also). I'm not 100% sure though and would need to experiment / write some tests to be sure but can we track in an issue for now

Reply:
#71